topology/probes/pmd/pmd.go

```diff
@@ -0,0 +1,122 @@
+package pmd
+
+import (
+        "time"
+        "os/exec"
+        "sync/atomic"
+        "sync"
+        "github.com/skydive-project/skydive/common"
+        "github.com/skydive-project/skydive/logging"
+        "github.com/skydive-project/skydive/topology/graph"
+)
+
+type PMDProbe struct {
+        graph.DefaultGraphListener
+        quit      chan bool
+        graph     *graph.Graph
+        state        int64
+        wg        sync.WaitGroup
+}
+func (o *PMDProbe)  OnNodeAdded(n *graph.Node) {
+        if tp, _ := n.GetFieldString("Name"); tp == "ovs-system" {
+                atomic.StoreInt64(&o.state, common.RunningState)
+                o.wg.Done()
+        }
+}
+
+
+func (o *PMDProbe) CoverageUpdate() error {
+        o.graph.Lock()
+        defer o.graph.Unlock()
+        commandLine := []string{"ovs-appctl","coverage/show"}
+        lines, err := launchOnSwitch(commandLine)
+        if err != nil {
+                logging.GetLogger().Infof("coverage/show failed with %s %s: ",  lines, err.Error())
+                return err
+        }
+        node := o.graph.LookupFirstNode(graph.Metadata{"Name": "ovs-system"})
+        if node != nil {
+                o.graph.AddMetadata(node, "coverage", lines)
+        }
+        intfs := o.graph.GetNodes(graph.Metadata{
+                "Type":    "ovsbridge",
+                "Datapath Type": "netdev",
+        })
+        if len(intfs) != 0 {
+                o.PMDUpdate()
+        }
+        return nil
+}
+
+func (o *PMDProbe) PMDUpdate() error {
+        commandLine := []string{"ovs-appctl", "dpif-netdev/pmd-stats-show"}
+        lines, err := launchOnSwitch(commandLine)
+        if err != nil {
+                logging.GetLogger().Infof("pmd-stats-show failed with %s %s: ",  lines, err.Error())
+                return err
+        }
+        node := o.graph.LookupFirstNode(graph.Metadata{"Name": "ovs-system"})
+        if node != nil {
+                o.graph.AddMetadata(node, "pmd", lines)
+        }
+        return nil
+}
+
+
+
+func (m *PMDProbe) Start() {
+        m.graph.AddEventListener(m)
+        m.wg.Add(1)
+Ready:  
+        for {   
+                switch atomic.LoadInt64(&m.state) {
```

Comment:
what about calling start from the event ? instead of looping which without sleep will consume CPU

Reply:
Thanks for the comment, will rewrite in the way suggested.  
OnNodeAdded was blocked when o.graph.Lock() was invoked directly in it, was not aware of it, hence the current commit was kind of mess, trying to work around the issue.
Now moved o.graph.Lock() to a goroutine, to be called after OnNodeAdded  is done, avoiding the issue.
Kind of like:
```
func (o *PMDProbe)  OnNodeAdded(n *graph.Node) {
                o.OvsStatsUpdate()
}


func (o *PMDProbe) OvsStatsUpdate() error {
        go func() {
                ticker := time.NewTicker(time.Duration(10) * time.Second)
                defer ticker.Stop()

                for {   
                        select {
                        case <-o.quit:
                               return
                        case <-ticker.C:
                               o.CoverageUpdate()
                        }
                }
        }()

        return nil
}
func (o *PMDProbe) CoverageUpdate() {
        o.graph.Lock()
        defer o.graph.Unlock()
}

```